docs/Security-Focused-Guide-for-AI-Code-Assistant-Instructions.md

```diff
@@ -1,64 +1,181 @@
+* **Use Industry Standards as Keywords:** Mention standards like **OWASP Top 10**, **OWASP ASVS**, **CWE/SANS Top 25**, or language-specific guidelines (Java's SEI CERT secure coding standards, for instance). For example: *"Code suggestions should adhere to OWASP Top 10 principles (e.g., avoid injection, enforce access control) and follow the OWASP ASVS requirements where applicable"*. [[42]](#42)
+* **Reference SAFECode and Other Guides:** SAFECode (Software Assurance Forum for Excellence in Code) publishes **Fundamental Practices for Secure Software Development**, and including a note about it can be beneficial. *"Our project follows SAFECode's secure development practices – the AI should prioritize those (e.g., proper validation, authentication, cryptography usage per SAFECode guidance)"*. This ties the assistant's suggestions to a comprehensive set of secure development principles. Similarly, you can reference the **Secure Software Development Lifecycle (SSDLC)** or standards like **ISO 27034** if your organization uses them, to align AI outputs with those processes. [[43]](#43)
+* **Compliance and Regulatory Security:** If you operate in an industry with specific regulations (like healthcare or finance), instruct the AI with those in mind. For instance: *"When generating code, consider compliance requirements (e.g., HIPAA privacy rules for medical data, PCI-DSS for credit card info) – do not output code that logs or transmits sensitive data in insecure ways"*. This ensures the AI's suggestions are compliant with external standards, for example, it might suggest encryption for personal data fields by default. [[44]](#44)
+* **Continuous Improvement Hooks:** Encourage practices that integrate with ongoing security processes. For example: *"Include comments or TODOs in code suggesting security reviews for complex logic, and note if any third-party component might need a future update or audit".* This way, the AI not only writes code but also flags areas for human scrutiny or future action. If the AI outputs a piece of cryptography code, for instance, it might add a comment "// TODO: security review for cryptographic implementation" because you instructed it that critical code should not go un-reviewed. [[45]](#45)
+* **Integrate and Act on Automated Security Tooling:** Instruct the AI to proactively run automated security tools (e.g., SAST, DAST, SCA) during code generation and refinement. Example: "When writing or reviewing code, run or simulate the use of tools like CodeQL, Bandit, Semgrep, or OWASP Dependency-Check. Identify any flagged vulnerabilities or outdated dependencies and revise the code accordingly. Repeat this process until the code passes all simulated scans". This positions the AI not just as a code generator but as a security-aware assistant capable of iterating based on tool feedback. [[46]](#46) [[47]](#47)
 
 ---
+
+## Appendix: Citations and References
+
```

Comment:
Personally I'd recommend using author-date (year) as the IDs, as it's easier to figure out what is what if you already know the key citations. But this works, so carry on!

Reply:
Agreed. Easier to find and also to add a new footnote `n` with manually numbered links will mean updating every footnote with the original `n`, `n+1`, `n+2` ... `m` with `m` being the last numbered footnote.


Here's an example of what's used in the C & C++ doc to achieve automatic footnotes `[^Cimpanu2019]`